src/macaron/build_spec_generator/build_command_patcher.py

```diff
@@ -0,0 +1,137 @@
+
+import logging
+from collections.abc import Mapping, Sequence
+
+from macaron.build_spec_generator.cli_command_parser import CLICommand, CLICommandParser, PatchCommandBuildTool
+from macaron.build_spec_generator.cli_command_parser.gradle_cli_parser import (
+    GradleCLICommandParser,
+    GradleOptionPatchValueType,
+)
+from macaron.build_spec_generator.cli_command_parser.maven_cli_parser import (
+    CommandLineParseError,
+    MavenCLICommandParser,
+    MavenOptionPatchValueType,
+    PatchBuildCommandError,
+)
+from macaron.build_spec_generator.cli_command_parser.unparsed_cli_command import UnparsedCLICommand
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+MVN_CLI_PARSER = MavenCLICommandParser()
+GRADLE_CLI_PARSER = GradleCLICommandParser()
+
+PatchValueType = GradleOptionPatchValueType | MavenOptionPatchValueType
+
+
+def _patch_commands(
+    cmds_sequence: Sequence[list[str]],
+    cli_parsers: Sequence[CLICommandParser],
+    patches: Mapping[
+        PatchCommandBuildTool,
+        Mapping[str, PatchValueType | None],
+    ],
+) -> list[CLICommand] | None:
+    """Patch the sequence of build commands, using the provided CLICommandParser instances.
+
+    For each command in `cmds_sequence`, it will be checked against all CLICommandParser instances until there is
+    one that can parse it, then a patch from ``patches`` is applied for this command if provided.
+
+    If a command doesn't have any corresponding ``CLICommandParser`` instance it will be parsed as UnparsedCLICommand,
+    which just holds the original command as a list of string, without any changes.
+    """
+    result: list[CLICommand] = []
+    for cmds in cmds_sequence:
+        effective_cli_parser = None
+        for cli_parser in cli_parsers:
+            if cli_parser.is_build_tool(cmds[0]):
+                effective_cli_parser = cli_parser
+                break
+
+        if not effective_cli_parser:
+            result.append(UnparsedCLICommand(original_cmds=cmds))
+            continue
+
+        try:
+            cli_command = effective_cli_parser.parse(cmds)
+        except CommandLineParseError as error:
+            logger.error(
+                "Failed to parse the cli command %s. Error %s.",
+                " ".join(cmds),
+                error,
+            )
+            return None
+
+        patch = patches.get(effective_cli_parser.build_tool, None)
+        if not patch:
+            result.append(cli_command)
+            continue
+
+        try:
+            new_cli_command = effective_cli_parser.apply_patch(
+                cli_command=cli_command,
+                options_patch=patch,
+            )
+        except PatchBuildCommandError as error:
+            logger.error(
+                "Failed to patch the mvn command %s. Error %s.",
+                " ".join(cmds),
+                error,
+            )
+            return None
+
+        result.append(new_cli_command)
+
+    return result
+
+
+def patch_commands(
```

Comment:
I'm not sure I see the value of having the two separated functions here. The public function only seems to pass the parsers and perform a small extra step on the results. I'd suggest combining them, making the parsers arg override the default list, if provided.

Reply:
I wonder if this align with what you have in mind ?
```
def patch_commands(
    cmds_sequence: Sequence[list[str]],
    patches: Mapping[
        PatchCommandBuildTool,
        Mapping[str, PatchValueType | None],
    ],
    cli_parsers: Sequence[CLICommandParser] | None = None,
) -> list[list[str]] | None:
    effective_cli_parsers = cli_parsers if cli_parsers else [MVN_CLI_PARSER, GRADLE_CLI_PARSER]
    # The body of _patch_commands using effective_cli_parsers  
```